src/Functions/FunctionsCharsetClassification.cpp

```diff
@@ -0,0 +1,247 @@
+        std::unordered_map<UInt16, Float64>& model)
+    {
+
+        const char * start = data;
+        const char * end = data + size;
+        CodePoint cp[simultaneously_codepoints_num] = {};
+        /// read_code_points returns the position of cp where it stopped reading codepoints.
+        size_t found = read_code_points(cp, start, end);
+        /// We need to start for the first time here, because first N - 1 codepoints mean nothing.
+        size_t i = N - 1;
+        size_t len = 0;
+        do
+        {
+            for (; i + N <= found; ++i)
+            {
+                UInt32 hash = 0;
+                for (size_t j = 0; j < N; ++j)
+                {
+                    hash <<= 8;
+                    hash += *(cp + i + j);
+                }
+                if (model[hash] == 0)
+                {
+                    model[hash] = 1;
+                    ++len;
+                }
+                ++model[hash];
+            }
+            i = 0;
+        } while (start < end && (found = read_code_points(cp, start, end)));
+
+        return len;
+    }
+
+
+    static void constant(String data, String & res)
+    {
+        const auto & encodings_freq = FrequencyHolder::getInstance().getEncodingsFrequency();
+
+        std::unordered_map<UInt16, Float64> model;
+        calculateStats(data.data(), data.size(), readCodePoints, model);
+
+        Float64 max_result = log(zero_frequency) * (max_string_size);
+        String poss_ans;
+        /// Go through the dictionary and find the charset with the highest weight
+        for (auto& item : encodings_freq)
+        {
+            Float64 score = naiveBayes(item.map, model, max_result);
+            if (max_result < score)
+            {
+                poss_ans = item.name;
+                max_result = score;
+            }
+        }
+
+        /* In our dictionary we have lines with form: <Language>_<Charset>
+         * If we need to find language of data, we return <Language>
+         * If we need to find charset of data, we return <Charset>.
+         */
+
+        size_t sep = poss_ans.find('_');
+        if (detect_language)
+        {
+            res = poss_ans.erase(0, sep + 1);
+        }
+        else
+        {
+            res = poss_ans.erase(sep, poss_ans.size() - sep);
+        }
+    }
+
+
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const auto & encodings_freq = FrequencyHolder::getInstance().getEncodingsFrequency();
+
+        res_data.reserve(1024);
+        res_offsets.resize(offsets.size());
+
+        size_t prev_offset = 0;
+        size_t res_offset = 0;
+
+        for (size_t i = 0; i < offsets.size(); ++i)
+        {
+            const char * haystack = reinterpret_cast<const char *>(&data[prev_offset]);
```

Comment:
Hm, it's bad that we ignore offsets, but just copy string as C-style.
And `reinterpret_cast` is not very good here.

Look at this example
https://github.com/ClickHouse/ClickHouse/blob/515cc74530d11e1b2b18a63141b66a15b94748ba/src/Columns/ColumnString.h#L98-L102